lib/pbio/src/geometry.c

```diff
@@ -0,0 +1,104 @@
+ *
+ * @param [in]  a       The first vector.
+ * @param [in]  b       The first vector.
+ * @param [out] output  The result.
+ */
+void pbio_geometry_vector_cross_product(pbio_geometry_xyz_t *a, pbio_geometry_xyz_t *b, pbio_geometry_xyz_t *output) {
+    output->x = a->y * b->z - a->z * b->y;
+    output->y = a->z * b->x - a->x * b->z;
+    output->z = a->x * b->y - a->y * b->x;
+}
+
+/**
+ * Gets the scalar projection of one vector onto the line spanned by another.
+ *
+ * @param [in]  input   The input vector.
+ * @param [in]  axis    The axis on which to project.
+ * @return              Signed projection.
+ */
+float pbio_geometry_vector_project(pbio_geometry_xyz_t *axis, pbio_geometry_xyz_t *input) {
+
+    // Normalize the given axis so its magnitude does not matter.
+    pbio_geometry_xyz_t unit_axis;
+    pbio_geometry_vector_normalize(axis, &unit_axis);
+
+    return unit_axis.x * input->x + unit_axis.y * input->y + unit_axis.z * input->z;
+}
+
+/**
+ * Maps the input vector using: output = map * input
+ *
+ * @param [in]  map     The 3x3 rotation matrix.
+ * @param [in]  input   The input vector
+ * @param [out] output  The resulting output vector.
+ */
+void pbio_geometry_vector_map(pbio_geometry_matrix_3x3_t *map, pbio_geometry_xyz_t *input, pbio_geometry_xyz_t *output) {
+    output->x = input->x * map->m11 + input->y * map->m12 + input->z * map->m13;
+    output->y = input->x * map->m21 + input->y * map->m22 + input->z * map->m23;
+    output->z = input->x * map->m31 + input->y * map->m32 + input->z * map->m33;
+}
+
+/**
+ * Gets a mapping (a rotation matrix) from two orthogonal base axes.
+ *
+ * @param [in]  x_axis  The X axis. Need not be normalized.
+ * @param [in]  z_axis  The Y axis. Need not be normalized.
+ * @param [out] map     The completed map, including the computed y_axis.
+ */
+pbio_error_t pbio_geometry_map_from_base_axes(pbio_geometry_xyz_t *x_axis, pbio_geometry_xyz_t *z_axis, pbio_geometry_matrix_3x3_t *map) {
+
+    pbio_geometry_xyz_t x_axis_normal;
+    pbio_geometry_vector_normalize(x_axis, &x_axis_normal);
+
+    pbio_geometry_xyz_t z_axis_normal;
+    pbio_geometry_vector_normalize(z_axis, &z_axis_normal);
+
+    // Assert that X and Z are orthogonal.
+    float inner_product = x_axis_normal.x * z_axis_normal.x + x_axis_normal.y * z_axis_normal.y + x_axis_normal.z * z_axis_normal.z;
+    if (inner_product > 0.001f || inner_product < -0.001f) {
+        return PBIO_ERROR_INVALID_ARG;
+    }
+
+    pbio_geometry_xyz_t y_axis_normal;
+    pbio_geometry_vector_cross_product(&z_axis_normal, &x_axis_normal, &y_axis_normal);
+
+    // Build the resulting matrix as [x_axis y_axis z_axis]
+    *map = (pbio_geometry_matrix_3x3_t) {
+        .m11 = x_axis_normal.x, .m12 = y_axis_normal.x, .m13 = z_axis_normal.x,
+        .m21 = x_axis_normal.y, .m22 = y_axis_normal.y, .m23 = z_axis_normal.y,
+        .m31 = x_axis_normal.z, .m32 = y_axis_normal.z, .m33 = z_axis_normal.z,
+    };
```

Comment:
For some cases, I like the notation above for clarity, because you can see at a glance that none of the assigned values depend on any existing values/state.

I was hoping the compiler was smart enough to avoid a memcpy.